Shipping labels: State machine

0nko opened the pull request:
This is a first split of #3224 (it's too big) and contains the creation workflow state machine.

I've chosen the state machine flow control approach to decrease the complexity of state management of the shipping label creation feature. The `ShippingLabelStateMachine` captures the main workflow screen + some of the details states of origin and shipping address verification. 

The side effects are exposed using the [Kotlin Flow](https://developer.android.com/kotlin/flow), which allows the ViewModel to asynchronously observe the `effects` property of the state machine.

This is still a WIP progress and additional states, events and side effects will be added for packaging, customs, shipping carrier and payment steps.

This PR adds the [state machine library dependency from Tinder](https://github.com/Tinder/StateMachine), which offers a simple DSL for defining the state machine behavior.

0nko committed:
Add the creation flow state machine

(cherry picked from commit 205be9be57f4b9569b67b543e3d1d04e94d6da8f)

0nko committed:
Add the shipping label creation state machine

0nko committed:
Remote the old class

0nko committed:
Fix the ktlint issues

0nko committed:
Merge branch 'feature/shipping-label-creation' into issue/3016-state-machine

nbradbury commented:
@0nko I took a first pass at this and it looks good, but I have no experience with Kotlin Flow and am nervous to merge this without another pair of eyes. I'll ask on Slack if anyone has Flow experience.

AmandaRiu reviewed (commented):
I read up through the Kotlin flow docs and Tinder diagram and this all _looks_ good, but it's a bit hard to really wrap my head around it at this level as it's all still very nebulous without anything interacting with the state machine. The one thing I did notice is we'll need to make sure we are f[ollowing the licensing requirements](https://github.com/Tinder/StateMachine/blob/master/LICENSE) for the Tinder library.

AmandaRiu commented:
@0nko Maybe it'd be helpful to add some unit tests for the state machine so we can follow what is happening for the purpose of the PR review?

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelStateMachine.kt:
```diff
@@ -0,0 +1,343 @@
+            }
+        }
+
+        state<DataLoading> {
+            on<DataLoaded> { event ->
+                val data = Data(event.originAddress, event.shippingAddress, setOf(ORIGIN_ADDRESS))
+                transitionTo(WaitingForUser(data), UpdateViewState(data))
+            }
+            on<DataLoadingFailed> {
+                transitionTo(DataLoadingFailure, ShowError(DataLoadingError))
+            }
+        }
```
A minor nitpick, don't you think removing the actual child classes imports, and scoping each class to its parent will make the code easier to read?
Like this:
```kotlin
        state<State.Idle> {
            on<Event.FlowStarted> { event ->
                transitionTo(State.DataLoading, SideEffect.LoadData(event.orderId))
            }
        }
```

0nko replied:
Perhaps. I'll refactor them.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelStateMachine.kt:
```diff
@@ -0,0 +1,343 @@
+
+        onTransition { transition ->
+            if (transition is StateMachine.Transition.Valid) {
+                Log.d(TAG, transition.toState.toString())
```
I think using WooLog here is better, it would help analyzing the app logs if someone complains about something, WDYT?

0nko replied:
Yup 👍

hichamboushaba reviewed (dismissed):
LGTM, 👍 for using StateFlow here.
I left a two nitpick comments 🙂

0nko committed:
Use parent class names when referencing events, states and side effects

0nko committed:
Use WooLog instead of Log

0nko committed:
Rename the state machine class

0nko committed:
Remove unused tag property

0nko committed:
Update the waiting state name

0nko committed:
Add unit test to test the initial loading sequence

0nko committed:
Add address verification unit test

0nko committed:
Add comments to the state machine

0nko commented:
> Maybe it'd be helpful to add some unit tests for the state machine so we can follow what is happening for the purpose of the PR review?

I added a couple of unit tests in `ShippingLabelsStateMachineTest` to illustrate the state machine flow. I also added documentation comments to the state machine class.

> The one thing I did notice is we'll need to make sure we are following the licensing requirements for the Tinder library.

It's a BSD-based license.

![image](https://user-images.githubusercontent.com/1522856/100863894-85883600-3495-11eb-8934-c81f50fc27cd.png)

As @hichamboushaba mentioned on Slack, we already use a library (Shimmer) with [the same license.](https://github.com/facebook/shimmer-android/blob/master/LICENSE)

@nbradbury @AmandaRiu Ready for another round. If you want to test the behavior with the actual UI, it's integrated in PR #3224.

0nko committed:
Merge branch 'develop' into issue/3016-state-machine

hichamboushaba commented:
@0nko just a reminder 🙂, before merging the PR and to align with the license requirement, we need to add the attribution to the list of licenses here: https://github.com/woocommerce/woocommerce-android/blob/develop/WooCommerce/src/main/res/raw/licenses.html

0nko commented:
@hichamboushaba Good point, thanks! :)

0nko committed:
Add the StateMachine library attributaion

hichamboushaba commented:
@0nko kudos for the state machine diagram, it makes understanding what's going a lot easier 👏 

One last thing, the licenses.html file is localized in the app, do you know what's the process to localize it? and if it should be done in this PR or later?

0nko commented:
@hichamboushaba Thanks! The localization is done later using GlotPress.

0nko committed:
Fix the state machine diagram

nbradbury reviewed (approved):
Awesome work on this! Sorry it took awhile for us to merge it, but we're all new to state machine and flow and needed more eyes on it. I'm good to :shipit:

0nko commented:
Thanks everyone! 🥳
